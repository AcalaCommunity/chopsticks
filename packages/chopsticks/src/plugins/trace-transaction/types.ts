import { HexString } from '@polkadot/util/types'
import { Registry } from '@polkadot/types/types'

export type CallTrace = {
  type: 'CALL' | 'CALLCODE' | 'STATICCALL' | 'DELEGATECALL' | 'CREATE' | 'SUICIDE'
  from: HexString
  to: HexString
  input: HexString
  value: HexString
  gas: number
  gasUsed: number
  output: HexString | null
  error: string | null
  revertReason: string | null
  depth: number
  calls: CallTrace[]
}

export type Step = {
  op: number
  pc: number
  depth: number
  gas: number
  stack: HexString[]
  memory: string[] | null
}

export type TraceOutcome =
  | {
      steps: Step[]
    }
  | {
      calls: CallTrace[]
    }

export const registerTypes = (registry: Registry) => {
  registry.register({
    Step: {
      op: 'u8',
      pc: 'Compact<u32>',
      depth: 'Compact<u32>',
      gas: 'Compact<u64>',
      stack: 'Vec<Bytes>',
      memory: 'Option<Vec<Bytes>>',
    },
    CallType: {
      _enum: {
        CALL: null,
        CALLCODE: null,
        STATICCALL: null,
        DELEGATECALL: null,
        CREATE: null,
        SUICIDE: null,
      },
    },
    CallTrace: {
      type: 'CallType',
      from: 'H160',
      to: 'H160',
      input: 'Bytes',
      value: 'U256',
      gas: 'Compact<u64>',
      gasUsed: 'Compact<u64>',
      output: 'Option<Bytes>',
      error: 'Option<String>',
      revertReason: 'Option<String>',
      depth: 'Compact<u32>',
      calls: 'Vec<CallTrace>',
    },
    TraceOutcome: {
      _enum: {
        Calls: 'Vec<CallTrace>',
        Steps: 'Vec<Step>',
      },
    },
    OpcodeConfig: {
      page: 'u32',
      pageSize: 'u32',
      disableStack: 'bool',
      enableMemory: 'bool',
    },
    TracerConfig: {
      _enum: {
        CallTracer: null,
        OpcodeTracer: 'OpcodeConfig',
      },
    },
  })
}
