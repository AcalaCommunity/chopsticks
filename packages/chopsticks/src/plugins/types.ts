export { rpc as newBlock, NewBlockParams } from './new-block'
