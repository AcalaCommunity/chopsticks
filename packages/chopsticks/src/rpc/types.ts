export { ChainProperties, RuntimeVersion } from '@acala-network/chopsticks-core'

export { default as author, AuthorHandlers } from './substrate/author'
export { handlers as chain, ChainHandlers, ChainHandlersAlias } from './substrate/chain'
export { default as payment, PaymentHandlers } from './substrate/payment'
export { default as state, StateHandlers } from './substrate/state'
export { default as system, SystemHandlers } from './substrate/system'
