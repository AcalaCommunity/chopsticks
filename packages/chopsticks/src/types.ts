/**
 * Chopsticks JSON RPC and CLI.
 *
 * @remarks
 * This package extends the `@acala-network/chopsticks-core` package a with JSON RPC and CLI support.
 *
 * @privateRemarks
 * Above is the package documentation for 'chopsticks' package.
 * `export` below is for tsdoc.
 *
 * @packageDocumentation
 */

export * from './plugins/types'
export { Context } from './rpc/shared'
