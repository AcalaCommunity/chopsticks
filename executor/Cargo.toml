[package]
name = "chopsticks-executor"
version = "0.0.0" # DO NOT CHANGE
description = "Chopsticks executor"
repository = "https://github.com/AcalaNetwork/chopsticks"
license = "Apache-2.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
crate-type = ["cdylib"]

[dependencies]
arrayvec = "0.7"
getrandom = { version = "0.2", default-features = false, features = ["js"] }
serde = { version = "1.0", default-features = false }
serde_json = { version = "1.0", default-features = false }
hex = { version = "0.4.3", default-features = false }
hex-literal = "0.4.1"
wasm-bindgen = "0.2.87"
serde-wasm-bindgen = "0.6.3"
wasm-bindgen-futures = "0.4"
js-sys = "0.3"
log = "0.4"
console_error_panic_hook = "0.1"
console_log = { version = "1.0" }

smoldot = { path = '../vendor/smoldot/lib', default-features = false }
smoldot-light = { path = '../vendor/smoldot/light-base', default-features = false }
no-std-net = { version = "0.6.0", default-features = false }
futures-lite = { version = "2.0.0", default-features = false, features = ["alloc"]}
derive_more = "0.99.17"
event-listener = { version = "4.0.2", default-features = false }
nom = { version = "7.1.3", features = [] }
slab = { version = "0.4.9", features = [] }
fnv = { version = "1.0.7", default-features = false }
hashbrown = { version = "0.14.2", default-features = false }
pin-project = "1.1.3"
async-channel = { version = "2.1.1", default-features = false }

[features]
default = []
std = ["smoldot/std", "smoldot-light/std"]

[profile.release]
codegen-units = 1
lto = true
