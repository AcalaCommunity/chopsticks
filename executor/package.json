{
	"name": "@acala-network/chopsticks-executor",
	"description": "Chopsticks executor",
	"version": "0.7.4-0",
	"license": "Apache-2.0",
	"repository": {
		"type": "git",
		"url": "https://github.com/AcalaNetwork/chopsticks"
	},
	"scripts": {
		"clean": "rm -rf web node",
		"build": "wasm-pack build --target web --out-dir web; wasm-pack build --target nodejs --out-dir node"
	},
	"files": [
		"web/chopsticks_*",
		"node/chopsticks_*"
	],
	"main": "node/chopsticks_executor.js",
	"types": "web/chopsticks_executor.d.ts",
	"module": "web/chopsticks_executor.js",
	"stableVersion": "0.7.3"
}
