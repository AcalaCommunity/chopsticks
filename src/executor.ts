import { HexString } from '@polkadot/util/types'
import { WebSocket } from 'ws'
import { hexToString, hexToU8a } from '@polkadot/util'
global.WebSocket = WebSocket

import { compactHex } from './utils'
import { get_metadata, get_runtime_version, run_task } from '../executor/pkg'

export type RuntimeVersion = {
  specName: string
  implName: string
  authoringVersion: number
  specVersion: number
  implVersion: number
  apis: [HexString, number][]
  transactionVersion: number
  stateVersion: number
}

export const getRuntimeVersion = async (code: HexString): Promise<RuntimeVersion> => {
  return get_runtime_version(code).then((version) => {
    version.specName = hexToString(version.specName)
    version.implName = hexToString(version.implName)
    return version
  })
}

export const getMetadata = async (code: HexString): Promise<HexString> => {
  return compactHex(hexToU8a(await get_metadata(code)))
}

export { run_task as runTask }
