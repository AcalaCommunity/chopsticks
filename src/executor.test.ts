import { HexString } from '@polkadot/util/types'
import { Metadata } from '@polkadot/types/metadata'
import { TypeRegistry } from '@polkadot/types'
import { decorateStorage } from '@polkadot/types/metadata/decorate'
import { describe, expect, it } from 'vitest'
import { getMetadata, getRuntimeVersion } from './executor'
import { readFileSync } from 'fs'
import path from 'path'

describe('wasm', () => {
  it('get metadata from wasm runtime', async () => {
    const code = String(readFileSync(path.join(__dirname, 'runtime.example'))).trim()
    expect(code.length).toBeGreaterThan(2)
    const registry = new TypeRegistry()
    const metadata = new Metadata(registry, await getMetadata(code as HexString))
    registry.setMetadata(metadata)
    expect(registry.metadata.pallets.length).toBeGreaterThan(0)
    const storage = decorateStorage(registry, metadata.asLatest, metadata.version)
    expect(storage.system).toBeTruthy
  })

  it('get runtime version from wasm runtime', async () => {
    const code = String(readFileSync(path.join(__dirname, 'runtime.example'))).trim()
    expect(code.length).toBeGreaterThan(2)

    const expectedRuntimeVersion = {
      specName: 'acala',
      implName: 'acala',
      authoringVersion: 1,
      specVersion: 2000,
      implVersion: 0,
      apis: [
        ['0xdf6acb689907609b', 3],
        ['0x37e397fc7c91f5e4', 1],
        ['0x40fe3ad401f8959a', 5],
        ['0xd2bc9897eed08f15', 3],
        ['0xf78b278be53f454c', 2],
        ['0xdd718d5cc53262d4', 1],
        ['0xab3c0572291feb8b', 1],
        ['0xbc9d89904f5b923f', 1],
        ['0x37c8bb1350a9a2a8', 1],
        ['0x6ef953004ba30e59', 1],
        ['0xf485c9145d3f0aad', 1],
        ['0xe3df3f2aa8a5cc57', 1],
        ['0xea93e3f16f3d6962', 1],
      ],
      transactionVersion: 1,
      stateVersion: 0,
    }

    expect(await getRuntimeVersion(code as HexString)).toMatchObject(expectedRuntimeVersion)
  })
})
