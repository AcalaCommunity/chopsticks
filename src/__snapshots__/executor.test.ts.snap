// Vitest Snapshot v1

exports[`wasm > decode & create proof works 1`] = `
{
  "0x06de3d8a54d27e44a9d5ce189618f22db4b49d95320d9021994c850f25b8e385": "0x0000a000005000000a00000000c8000000c800000a0000000a00000040380000580200000000500000c8000000e87648170000001e0000000000000000e8764817000000000000000000000000e87648170000000000000000000000e8030000009001001e00000000000000009001004038000000000000000000000a0000000a0000000a00000001000000010500000001c8000000060000005802000002000000580200000200000059000000000000001e0000002800000000c817a804000000000200000014000000",
  "0x1cb6f36e027abb2091cfb5110ab5087f06155b3cd9a8c9e5e9a23fd5dc13a5ed": "0x873c991000000000",
  "0x1cb6f36e027abb2091cfb5110ab5087f38316cbf8fa0da822a20ac1c55bf1be3": "0xcc15000000000000",
  "0x1cb6f36e027abb2091cfb5110ab5087f7a414cb008e0e61e46722aa60abdd672": "0xb46e626e564d1385d21257d8cb59236ca125657f2a14bebd1af174317ca1055c",
  "0x1cb6f36e027abb2091cfb5110ab5087f7ce678799d3eff024253b90e84927cc6": "0x13c6d5340826d75f6258aad1cf4478c20cae5bbaf189e2fe6e0c997aec0a8c3c",
  "0x1cb6f36e027abb2091cfb5110ab5087fd077dfdb8adb10f78f10a5df8742c545": "0x0159a565b0fb89d8eec07b522e4e30e8bc29489ef7a9609dfec01855bd125d2e33",
  "0x63f78c98723ddc9073523ef3beefda0c4d7fefc408aac59dbfe80a72ac8e3ce563f5a4efb16ffa83d0070000": "0xd205bfd64a59c64fe84480fda7dafd773cb029530c4efe8441bf1f4332bfa48a",
  "0x6a0da05ca59913bc38a8630590f2627c1d3719f5b0b12c7105c073c50744594863f5a4efb16ffa83d0070000": "0x20e8030000d4070000d6070000db070000dc070000f0070000f2070000f3070000",
  "0x6a0da05ca59913bc38a8630590f2627cf12b746dcf32e843354583c9702cc02063f5a4efb16ffa83d0070000": "0x20e8030000d4070000d6070000db070000dc070000f0070000f2070000f3070000",
  "0xcd710b30bd2eab0352ddcc26417aa1949e94c040f5e73d9b7addd6cb603d15d363f5a4efb16ffa83d0070000": undefined,
}
`;

exports[`wasm > decode & create proof works 2`] = `
{
  "chainAvailabilityPeriod": 10,
  "codeRetentionPeriod": 0,
  "disputeConclusionByTimeOutPeriod": 600,
  "disputeMaxSpamSlots": 2,
  "disputePeriod": 6,
  "disputePostConclusionAcceptancePeriod": 600,
  "groupRotationFrequency": 10,
  "hrmpChannelMaxCapacity": 102400,
  "hrmpChannelMaxMessageSize": 14400,
  "hrmpChannelMaxTotalSize": 30,
  "hrmpMaxMessageNumPerCandidate": 10,
  "hrmpMaxParachainInboundChannels": 0,
  "hrmpMaxParachainOutboundChannels": 30,
  "hrmpMaxParathreadInboundChannels": 102400,
  "hrmpMaxParathreadOutboundChannels": 0,
  "hrmpOpenRequestTtl": 1215752192,
  "hrmpRecipientDeposit": "0x000003e8000000000000000000000017",
  "hrmpSenderDeposit": "0x4876e800000000000000000000000017",
  "maxCodeSize": 10485760,
  "maxDownwardMessageSize": 51200,
  "maxHeadDataSize": 20480,
  "maxPovSize": 5242880,
  "maxUpwardMessageNumPerCandidate": 10,
  "maxUpwardMessageSize": 51200,
  "maxUpwardQueueCount": 10,
  "maxUpwardQueueSize": 51200,
  "maxValidators": 200,
  "maxValidatorsPerCore": null,
  "nDelayTranches": 89,
  "neededApprovals": 30,
  "noShowSlots": 2,
  "parathreadCores": 0,
  "parathreadRetries": 10,
  "preferredDispatchableUpwardMessagesStepWeight": 100000000000,
  "relayVrfModuloSamples": 40,
  "schedulingLookahead": 1281,
  "threadAvailabilityPeriod": 1,
  "validationUpgradeDelay": 600,
  "validationUpgradeFrequency": 14400,
  "zerothDelayTrancheWidth": 0,
}
`;

exports[`wasm > decode & create proof works 3`] = `"0xfc0b4018744f7cf8390204c6d2fbef8ede8c556942e5f5c63bd4f36075897bb3"`;

exports[`wasm > decode & create proof works 4`] = `
[
  "0x5703f5a4efb16ffa83d00700008420e8030000d4070000d6070000db070000dc070000f0070000f2070000f3070000",
  "0x5f0fb7edc4cbc65e03d0070000d40700005501e80300000090010000900100000000000000000001708391b220ce64bab1cc4f79559352e59882ee5bb58329085c2400079c56153a00e8764817000000000000000000000000e87648170000000000000000000000",
  "0x5e057a605f506cfcd4070000d00700005501e80300000090010000900100000000000000000001d3772d60305c2cbd7acf502a11881e0eef47fdfc85c7d2a0d6241e06ea4a971000e8764817000000000000000000000000e87648170000000000000000000000",
  "0x5ee678799d3eff024253b90e84927cc68013c6d5340826d75f6258aad1cf4478c20cae5bbaf189e2fe6e0c997aec0a8c3c",
  "0x8000c0787694c040f5e73d9b7addd6cb603d15d363f5a4efb16ffa83d00700000401485e4993f016e2d2f8e5f43be7bb2594860400",
  "0x80204080480c6b47252c3414177fcd8441cc689703241343e1ef84be0b206cc24b043f1c801f33c1df517954d9820ca15fa6afb41e88a4d9101e0839a6b162f4eeba54804a",
  "0x5e8b652b2ae6ed1ddc070000d00700005501e803000000900100009001000000000000000000010a492e67d36615052a89ba50498ed5cf0a587b31fd6f3da464e5d111e4e8a81c00e8764817000000000000000000000000e87648170000000000000000000000",
  "0x5f0077dfdb8adb10f78f10a5df8742c545840159a565b0fb89d8eec07b522e4e30e8bc29489ef7a9609dfec01855bd125d2e33",
  "0x8000c08090ebabb34893b8aef39ad17e906edea21d557366f6c759432622d651ef85d15880427ca6704596351b770d9c0a79343f2c1570d2bf146bf8dbd0115e4c853f3a4e",
  "0x5f050452a22bee61fad0070000f20700005501e80300000090010000900100000000000000000001e41d29d36ff0b7c6007ec17b68ba6165066170fc67eb9755fbd7ed60bef647c100e8764817000000000000000000000000e87648170000000000000000000000",
  "0x5e4f36708366b722d0070000e8030000d4e8030000009001000090010000000000000000000000e8764817000000000000000000000000e87648170000000000000000000000",
  "0x80004280df3c444cc602cfb8e35514a12c70fe2f5b2ec2683a3e19889c2bce42e8df4231808bbdb281c19f2e0405ebdcdd3eed583c83be8c5bf317070ab8e144c77aecab54",
  "0x9f0cb6f36e027abb2091cfb5110ab5087f9960685f06155b3cd9a8c9e5e9a23fd5dc13a5ed20873c991000000000685f08316cbf8fa0da822a20ac1c55bf1be320cc15000000000000505f0e7b9012096b41c4eb3aaf947f6ea42908000080950cc9876d11a25f6c56e833f048b7fc32caf0bfbb8d9ff4b6ea993f4e984698806edd811d66d803965303f11c4bce3a94b9af68520b9c75e3b60b46b9961c3932685f090e2fbf2d792cb324bffa9427fe1f0e20b94ccb00fa55cb00",
  "0x9f06de3d8a54d27e44a9d5ce189618f22d3008505f0e7b9012096b41c4eb3aaf947f6ea4290802004c5f03b4123b2e186e07fb7bad5dda5f55c0040080f71eda18869d6ad02c744d873d72c05d01fa458c4da7038805af33d407c65c7b",
  "0x5e36c65ca123d5fbd0070000dc0700005501e80300000090010000900100000000000000000001c5120071fe1663a5a21b96d87446ac903ef1bf87dd729b8b9e92ca42ca303e7400e8764817000000000000000000000000e87648170000000000000000000000",
  "0x5e4361975d97255ddb070000d00700005501e803000000900100009001000000000000000000018c74408c28f6627f58a8e10b3ec06b98b0166bfb148641a35c5cdfc7a56e68fb00e8764817000000000000000000000000e87648170000000000000000000000",
  "0x804390801737775bc254f18aa4a817286d58a84638bd49932cd69e763be2107bfd65a5c3803be255b7e517fdbbb01870006388d720dc86fab24cb17c563e80638471f3b7fd80ee94abc5b7c0aa800c057463ac4b0e4cee6bda525cfe5608717c8c5378ed053a80910c01e2f75ba76fa6a60c7cd8849c5c4745ce479199be8b04c32f17a4fc616680da68cf8effc6c5ec15c0f401c4ee20e662248c1c185495886f6f92b51744d3cc",
  "0x7f17f78c98723ddc9073523ef3beefda0c4d7fefc408aac59dbfe80a72ac8e3ce563f5a4efb16ffa83d007000080d205bfd64a59c64fe84480fda7dafd773cb029530c4efe8441bf1f4332bfa48a",
  "0x5ec648b30353eed1d0070000db0700005501e8030000009001000090010000000000000000000100382d804f7b02c95c2d6b3b6a9cad2c97c2d7570bdfd32f8e8d961754b74b6d00e8764817000000000000000000000000e87648170000000000000000000000",
  "0x5e36ad4314650419d0070000f00700005501e80300000090010000900100000000000000000001de18e43e692ffe15ea7a85abbce774e6944c56270a846777352827c7596f6c3e00e8764817000000000000000000000000e87648170000000000000000000000",
  "0x80104080a1fc2c779380c33b94b304c281e55ca5a0807804fe9530c22fdfa76628beb42680692755c9209155bd724071de137657f492de5bb5479738d1af7119739b4df473",
  "0x9f05207f03cfdce586301014700e2c25934040685f0d9ef3b78afddab7f5c7142131132ad42001000000000000004c5f0ec2d17a76153ff51817f12d9cfc3c7f0400",
  "0x9f012b746dcf32e843354583c9702cc020c01080bb7fd6b2fa22db69bca4f70304addc539798fc6174ddae0c5865c71ff96e53415c5700bd9a93e85e3ce1d20700002408d6070000dc0700004c57077a93d174890f1ff20700001404d0070000",
  "0x5e046fae65527199f2070000d0070000d4e8030000009001000090010000000000000000000000e8764817000000000000000000000000e87648170000000000000000000000",
  "0x5f09a99274c2ff3639d0070000d60700005501e80300000090010000900100000000000000000001d10c375b7f3756dd38a6fd0967fa4a92199606619c76c5aedc139eb584e8628b00e8764817000000000000000000000000e87648170000000000000000000000",
  "0x9f06604cff828a6e3f579ca6c59ace013d15af80501a70ba3d699bb9126e453b817d3594618486b7455fb052fab74ed939c8db6580a4c3594adfbea64ada3d4b236078605e76b6c08582af892ffe107e2be422384280d5f81eca86aee22ba30e880daf0795627f12862c94a6d6add61790d47cc4174780eaca9a51c780ffcf7f2075adef97b380408d38589e510e2a158e078b1cbd05a580e450c660302bd7ac657ccc169f4a3542dea29d4e3b07d14fe0565e790586ef2e808b324b7e3655d04c92725d3982a8fc54a6d7219215caaa1274c161d16a606ff780768c9728ddfa5bde58b60116f5a720e912ba2dad17b047967c6fcd63ee0e9aa080e83e651d1018ce1a41f9635bc29355efeeadf688ab261358f5d84630c184d22a80cd7728c9b53ec87bc6b1011aa31c30a4493a3282a701edcdfc3fd2460d16c936",
  "0x800410801eacd1970b27e7428c46fc7a1cd539390d664880f8665d1f5d9df711aef78ffc806dfac4a6f21c63e0cee19d347063fd8cdd56b9741d81c5842d27abed9542be5f",
  "0x9f0d710b30bd2eab0352ddcc26417aa19440824c5f03c716fb8fff3de61a883bb76adb34a20400807840ffa4227d602af6d5bd8e9a5c9abbf856ebd961515e561b9f8bd10877d31c4c5f0dad6eef5c4b1c68eaa71ea17a02d9de0400",
  "0x5e1c472775baca93f0070000d00700005501e803000000900100009001000000000000000000013b264b3f72ecfeddbdad3b409887206a18d387f1b81856c0383014c9f559992000e8764817000000000000000000000000e87648170000000000000000000000",
  "0x5e968c2de8299067f3070000d0070000d4e8030000009001000090010000000000000000000000e8764817000000000000000000000000e87648170000000000000000000000",
  "0x5e55c8e02d73966fd6070000d00700005501e803000000900100009001000000000000000000019a436c9ae233f8b3cc6a379dc35ff6a3c4153b22a96db9fc81c9cee506a252b900e8764817000000000000000000000000e87648170000000000000000000000",
  "0x9f0d3719f5b0b12c7105c073c507445948c01080bb7fd6b2fa22db69bca4f70304addc539798fc6174ddae0c5865c71ff96e53415c5700bd9a93e85e3ce1d20700002408d6070000dc0700004c57077a93d174890f1ff20700001404d0070000",
  "0x9e0da05ca59913bc38a8630590f2627c02888081036a00b8dc7fbcf94f6630e920116e4d272ddcceab72b3bf2ab4f420a361ae8094da7ace1d340bf65922ea519cc3ba35a47fbc93054b01e28976b253faf5a51780f47f9f2800119f22dbcd10050587ab9a8ced42ea869506c63ff6b62c7ba24252",
  "0x5f04b49d95320d9021994c850f25b8e385e9020000a000005000000a00000000c8000000c800000a0000000a00000040380000580200000000500000c8000000e87648170000001e0000000000000000e8764817000000000000000000000000e87648170000000000000000000000e8030000009001001e00000000000000009001004038000000000000000000000a0000000a0000000a00000001000000010500000001c8000000060000005802000002000000580200000200000059000000000000001e00000028000000",
  "0x5e414cb008e0e61e46722aa60abdd67280b46e626e564d1385d21257d8cb59236ca125657f2a14bebd1af174317ca1055c",
  "0x5e04d2a15ab51127e8030000d0070000d4e8030000009001000090010000000000000000000000e8764817000000000000000000000000e87648170000000000000000000000",
  "0x5e4be1ee506d55f8d0070000f3070000d4e8030000009001000090010000000000000000000000e8764817000000000000000000000000e87648170000000000000000000000",
  "0x80203080f12b3be1de5704b86affeaf4178a0ca8dea056893c12a03c3500ca8af78c1280802166c372b218427bae449c6e50a075b4bc604253cea4ad0ee8ee7c16b498f3e980750fa8bb7472874648dae3648edb7c4900fd5b779b1df2ed102069dfda793608",
  "0x800014808fc31cedc9f5a694fdde1823788928749f52e1c637dfbf315637fd89eace451c80b58f7e709e14c8ad32651a7b6f13b2ee8aa4a9566413b985e64eb5f3918a58e9",
  "0x800804802942f9dc1f55813e7f51d1e2e2144fe439074c4f5cdc26ecf1fe8339da8b3b1c80d0d1055f2e6a21fa35bf8b8f3811e00185a1b3564ce394e5f1004284c578a883",
]
`;

exports[`wasm > decode & create proof works 5`] = `
{
  "0x06de3d8a54d27e44a9d5ce189618f22db4b49d95320d9021994c850f25b8e385": "0x0000a000005000000a00000000c8000000c800000a0000000a00000040380000580200000000500000c8000000e87648170000001e0000000000000000e8764817000000000000000000000000e87648170000000000000000000000e8030000009001001e00000000000000009001004038000000000000000000000a0000000a0000000a00000001000000010500000001c8000000060000005802000002000000580200000200000059000000000000001e00000028000000",
  "0x1cb6f36e027abb2091cfb5110ab5087f06155b3cd9a8c9e5e9a23fd5dc13a5ed": "0x873c991000000000",
  "0x1cb6f36e027abb2091cfb5110ab5087f38316cbf8fa0da822a20ac1c55bf1be3": "0xcc15000000000000",
  "0x1cb6f36e027abb2091cfb5110ab5087f7a414cb008e0e61e46722aa60abdd672": "0xb46e626e564d1385d21257d8cb59236ca125657f2a14bebd1af174317ca1055c",
  "0x1cb6f36e027abb2091cfb5110ab5087f7ce678799d3eff024253b90e84927cc6": "0x13c6d5340826d75f6258aad1cf4478c20cae5bbaf189e2fe6e0c997aec0a8c3c",
  "0x1cb6f36e027abb2091cfb5110ab5087fd077dfdb8adb10f78f10a5df8742c545": "0x0159a565b0fb89d8eec07b522e4e30e8bc29489ef7a9609dfec01855bd125d2e33",
  "0x63f78c98723ddc9073523ef3beefda0c4d7fefc408aac59dbfe80a72ac8e3ce563f5a4efb16ffa83d0070000": "0xd205bfd64a59c64fe84480fda7dafd773cb029530c4efe8441bf1f4332bfa48a",
  "0x6a0da05ca59913bc38a8630590f2627c1d3719f5b0b12c7105c073c50744594863f5a4efb16ffa83d0070000": "0x20e8030000d4070000d6070000db070000dc070000f0070000f2070000f3070000",
  "0x6a0da05ca59913bc38a8630590f2627cf12b746dcf32e843354583c9702cc02063f5a4efb16ffa83d0070000": "0x20e8030000d4070000d6070000db070000dc070000f0070000f2070000f3070000",
  "0xcd710b30bd2eab0352ddcc26417aa1949e94c040f5e73d9b7addd6cb603d15d363f5a4efb16ffa83d0070000": "0x01",
}
`;
